File: new/Kotlin/home_accounting_query/src/main/kotlin/entities/Subcategories.kt
package com.sz.home_accounting.query.entities

import java.nio.ByteBuffer

enum class SubcategoryCode {
    Comb,
    Comc,
    Fuel,
    Prcn,
    Incc,
    Expc,
    Exch,
    Trfr,
    None
}

enum class SubcategoryOperationCode {
    Incm,
    Expn,
    Spcl
}

data class Subcategory(val id: Int,
                       val name: String,
                       val code: SubcategoryCode,
                       val operationCode: SubcategoryOperationCode,
                       val category: Int) {
    companion object {
        fun fromBinary(buffer: ByteBuffer): Map<Int, Subcategory> {
            var size = buffer.getShort()
            val result = mutableMapOf<Int, Subcategory>()
            while (size-- > 0) {
                val id = buffer.getShort().toInt()
                val subcategory = create(id, buffer)
                result[id] = subcategory
            }
            return result
        }

        private fun create(id: Int, buffer: ByteBuffer): Subcategory {
            val size = buffer.get().toInt()
            val array = ByteArray(size)
            buffer.get(array)
            val name = String(array)
            val code = SubcategoryCode.entries[buffer.get().toInt()]
            val operationCode = SubcategoryOperationCode.entries[buffer.get().toInt()]
            val category = buffer.getShort().toInt()
            return Subcategory(id, name, code, operationCode, category)
        }

        fun toBinary(subcategories: Map<Int, Subcategory>, buffer: ByteBuffer) {
            buffer.putShort(subcategories.size.toShort())
            for ((id,subcategory) in subcategories) {
                buffer.putShort(id.toShort())
                subcategory.toBinary(buffer)
            }
        }
    }

    private fun toBinary(buffer: ByteBuffer) {
        val bytes = name.toByteArray()
        buffer.put(bytes.size.toByte())
        buffer.put(bytes)
        buffer.put(code.ordinal.toByte())
        buffer.put(operationCode.ordinal.toByte())
        buffer.putShort(category.toShort())
    }
}

File: new/dotNET/HomeAccountingServiceClient/HomeAccountingServiceClientLibrary/entities/Subcategories.cs
namespace HomeAccountingServiceClientLibrary.entities;

public enum SubcategoryCode {
    Comb,
    Comc,
    Fuel,
    Prcn,
    Incc,
    Expc,
    Exch,
    Trfr,
    None
}

public enum SubcategoryOperationCode {
    Incm,
    Expn,
    Spcl
}

public record Subcategory(int Id, string Name, SubcategoryCode Code, SubcategoryOperationCode OperationCode, int Category)
{
    internal static Dictionary<int, Subcategory> FromBinary(BinaryReader reader)
    {
        var size = reader.ReadInt16();
        var result = new Dictionary<int, Subcategory>();
        while (size-- > 0)
        {
            var subcategory = Create(reader);
            result[subcategory.Id] = subcategory;
        }
        return result;
    }
    
    private static Subcategory Create(BinaryReader reader)
    {
        var id = (int)reader.ReadInt16();
        var nameSize = reader.ReadByte();
        var name = System.Text.Encoding.UTF8.GetString(reader.ReadBytes(nameSize));
        var code = (SubcategoryCode)reader.ReadByte();
        var operationCode = (SubcategoryOperationCode)reader.ReadByte();
        var category = (int)reader.ReadInt16();
        return new Subcategory(id, name, code, operationCode, category); 
    }
}